support-lib/jni/djinni_support.cpp

```diff
@@ -239,8 +239,13 @@ jint JniEnum::ordinal(JNIEnv * env, jobject obj) const {
 
 LocalRef<jobject> JniEnum::create(JNIEnv * env, jint value) const {
     LocalRef<jobject> values(env, env->CallStaticObjectMethod(m_clazz.get(), m_staticmethValues));
+    jniExceptionCheck(env);
     DJINNI_ASSERT(values, env);
-    return LocalRef<jobject>(env, env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()), value));
+    LocalRef<jobject> result(env,
+                             env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()),
+                                                        value));
+    jniExceptionCheck(env);
+    return std::move(result);
```

Comment:
Yeah, I get that.  That's not what I was wondering about.  What I wonder about is what thing the compiler might do in case 1 which would make it unsafe, which motivates the standard being defined that way.  I don't know of any meaningful optimizations which would make that unsafe.  The only ones I can think of (constant value propagation, or overlapping x with some other piece of data of the right value) would have to be disabled by RVO anywa (in the theoretical case where RVO could occur in such a case), since the function foo() has to put X in a location it can't control.  Or put a different way, if mutating a const value could be unsafe, why couldn't the compiler effectively create a non-const xtmp for RVO, and treat x like a const reference to it.  Seems an equally valid response to the situation, since all guaranteed behavior in the language (like const vs. non-const method overloading) isn't based on whether the variable "was created const" but only whether it is "accessed as const".

I might need a compiler expert to answer that, though.  Or it might just be that the standard specified the "same type" limitation specifically so the compiler didn't have to make potentially implementation-dependent decisions of which optimizations take precedence over which others.

The "same type" limitation makes me also wonder whether this might be a legitimate case for declaring a return type of "const X", though I still dislike that (as I similarly dislike arguments of type "const X") because it exposes implementation details to the caller.